File: Data Structures & Algorithms/balanced-binary-tree/submission-2.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public boolean isBalanced(TreeNode root) {
        boolean[] result = new boolean[1];
        result[0] = true;

        dfs(root, result);
        return result[0];
    }

    int dfs(TreeNode root, boolean[] result) {
        if (root == null) {
            return 0;
        }
        int leftHeight = dfs(root.left, result);
        int rightHeight = dfs(root.right, result);

        if (Math.abs(leftHeight - rightHeight) > 1) {
            result[0] = false;
        }

        return 1 + Math.max(leftHeight, rightHeight);
    }
}

File: Data Structures & Algorithms/binary-tree-maximum-path-sum/submission-2.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int maxPathSum(TreeNode* root) {
        maxPathOverall = INT_MIN;
        dfs(root);

        return maxPathOverall;
    }

private:
    int maxPathOverall;
    
    int dfs(TreeNode* root) {
        if (root == nullptr)
            return 0;

        int leftGain = dfs(root->left);
        int rightGain = dfs(root->right);

        int maxPathHere = root->val + max(0, leftGain) + max(0, rightGain);
        maxPathOverall = max(maxPathHere, maxPathOverall);

        return root->val + max({0, leftGain, rightGain});
    }
};
